diff --git a/drivers/usb/storage/usb.c b/drivers/usb/storage/usb.c
index 5c4fe07..816238c 100644
--- a/drivers/usb/storage/usb.c
+++ b/drivers/usb/storage/usb.c
@@ -1035,6 +1035,7 @@ static int storage_probe(struct usb_interface *intf,
 	struct us_data *us;
 	int result;
 	int size;
+	unsigned long timeout = (20 * HZ + INITIAL_JIFFIES);
 
 	/*
 	 * If the device isn't standard (is handled by a subdriver
@@ -1061,6 +1062,9 @@ static int storage_probe(struct usb_interface *intf,
 			id->idVendor, id->idProduct);
 	}
 
+	while(time_before(jiffies, timeout))
+		schedule();
+
 	result = usb_stor_probe1(&us, intf, id, unusual_dev);
 	if (result)
 		return result;
